// discord-trivia/dist/src/Classes/RootComponent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var RootComponent = /** @class */ (function () {
    function RootComponent(root) {
        var _this = this;
        this.reply = {
            message: function (options) {
                return _this.entity.reply(options);
            },
            interaction: function (options) {
                return _this.entity.reply(options);
            },
        };
        // The following error occures without ':any'
        // The inferred type of 'followUp' cannot be named without a reference to 'discord.js/node_modules/discord-api-types/v9'. This is likely not portable. A type annotation is necessary.
        this.followUp = {
            message: function (options) {
                return _this.entity.reply(options);
            },
            interaction: function (options) {
                return _this.entity.followUp(options);
            },
        };
        this.entity = root;
        if (root.content) {
            this.type = "message";
        }
        else if (root.applicationId) {
            this.type = "interaction";
        }
        else {
            throw new TypeError("root argument must be of type Message or CommandInteraction");
        }
        this.channel = root.channel;
        this.guild = root.guild;
        this.hostMember = root.member;
    }
    return RootComponent;
}());
exports.default = RootComponent;
