// discord-trivia/dist/src/Functions/prepareCustomQuestions.js
"use strict";
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
Object.defineProperty(exports, "__esModule", { value: true });
var open_trivia_db_1 = require("open-trivia-db");
/**
 * Takes an array of custom questions and parses them to be trivia game ready.
 * @param {CustomQuestion[]} questions - The custom questions.
 * @returns {Question[]} The parsed questions.
 */
function prepareCustomQuestions(questions) {
    return questions.map(function (q) {
        var value = q.value, difficulty = q.difficulty, correctAnswer = q.correctAnswer, incorrectAnswers = q.incorrectAnswers, category = q.category;
        var parsedQuestion = {
            value: value,
            category: new open_trivia_db_1.Category(category || "general knowledge").strictName,
            type: incorrectAnswers.length == 1 ? "boolean" : "multiple",
            difficulty: difficulty || "medium",
            correctAnswer: correctAnswer,
            incorrectAnswers: incorrectAnswers,
            allAnswers: open_trivia_db_1.OpenTDBUtil.shuffleArray(__spreadArray([
                correctAnswer
            ], incorrectAnswers, true)),
            checkAnswer: function (arg) {
                return arg.toLowerCase() == correctAnswer.toLowerCase();
            },
        };
        return parsedQuestion;
    });
}
exports.default = prepareCustomQuestions;
