// discord-trivia/dist/src/Components/messageActionRows.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Disabled = exports.buttonRowQueue = exports.buttonRowChoicesBoolean = exports.buttonRowChoicesMultiple = void 0;
var discord_js_1 = require("discord.js");
// The following errors occure without :any
// The inferred type of '...' cannot be named without a reference to 'discord.js/node_modules/discord-api-types/v9'. This is likely not portable. A type annotation is necessary.
var buttonRowChoicesMultiple = new discord_js_1.MessageActionRow().addComponents([
    new discord_js_1.MessageButton().setCustomId("0").setLabel("A").setStyle("PRIMARY"),
    new discord_js_1.MessageButton().setCustomId("1").setLabel("B").setStyle("PRIMARY"),
    new discord_js_1.MessageButton().setCustomId("2").setLabel("C").setStyle("PRIMARY"),
    new discord_js_1.MessageButton().setCustomId("3").setLabel("D").setStyle("PRIMARY"),
]);
exports.buttonRowChoicesMultiple = buttonRowChoicesMultiple;
var buttonRowChoicesBoolean = new discord_js_1.MessageActionRow().addComponents([
    new discord_js_1.MessageButton().setCustomId("1").setLabel("True").setStyle("PRIMARY"),
    new discord_js_1.MessageButton().setCustomId("0").setLabel("False").setStyle("DANGER"),
]);
exports.buttonRowChoicesBoolean = buttonRowChoicesBoolean;
var buttonRowQueue = new discord_js_1.MessageActionRow().addComponents([
    new discord_js_1.MessageButton().setCustomId("1").setLabel("Join").setStyle("SUCCESS"),
]);
exports.buttonRowQueue = buttonRowQueue;
var buttonRowChoicesMultipleDisabled = new discord_js_1.MessageActionRow().addComponents([
    new discord_js_1.MessageButton()
        .setCustomId("0")
        .setLabel("A")
        .setStyle("PRIMARY")
        .setDisabled(true),
    new discord_js_1.MessageButton()
        .setCustomId("1")
        .setLabel("B")
        .setStyle("PRIMARY")
        .setDisabled(true),
    new discord_js_1.MessageButton()
        .setCustomId("2")
        .setLabel("C")
        .setStyle("PRIMARY")
        .setDisabled(true),
    new discord_js_1.MessageButton()
        .setCustomId("3")
        .setLabel("D")
        .setStyle("PRIMARY")
        .setDisabled(true),
]);
var buttonRowChoicesBooleanDisabled = new discord_js_1.MessageActionRow().addComponents([
    new discord_js_1.MessageButton()
        .setCustomId("1")
        .setLabel("True")
        .setStyle("PRIMARY")
        .setDisabled(true),
    new discord_js_1.MessageButton()
        .setCustomId("0")
        .setLabel("False")
        .setStyle("DANGER")
        .setDisabled(true),
]);
var buttonRowQueueDisabled = new discord_js_1.MessageActionRow().addComponents([
    new discord_js_1.MessageButton()
        .setCustomId("1")
        .setLabel("Join")
        .setStyle("SUCCESS")
        .setDisabled(true),
]);
// The following error occures without :any
// The inferred type of 'Disabled' cannot be named without a reference to 'discord.js/node_modules/discord-api-types/v9'. This is likely not portable. A type annotation is necessary.
exports.Disabled = {
    buttonRowChoicesMultipleDisabled: buttonRowChoicesMultipleDisabled,
    buttonRowChoicesBooleanDisabled: buttonRowChoicesBooleanDisabled,
    buttonRowQueueDisabled: buttonRowQueueDisabled,
};
