// discord-trivia/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.prepareCustomQuestions = exports.TriviaManager = exports.TriviaGame = exports.TriviaCommandBuilder = exports.EmbedGenerator = exports.DiscordTriviaError = void 0;
// import CanvasGenerator from "./src/Classes/CanvasGenerator";
var DiscordTriviaError_1 = require("./src/Classes/DiscordTriviaError");
exports.DiscordTriviaError = DiscordTriviaError_1.default;
var EmbedGenerator_1 = require("./src/Classes/EmbedGenerator");
exports.EmbedGenerator = EmbedGenerator_1.default;
var TriviaCommandBuilder_1 = require("./src/Classes/TriviaCommandBuilder");
exports.TriviaCommandBuilder = TriviaCommandBuilder_1.default;
var TriviaGame_1 = require("./src/Classes/TriviaGame");
exports.TriviaGame = TriviaGame_1.default;
var TriviaManager_1 = require("./src/Classes/TriviaManager");
exports.TriviaManager = TriviaManager_1.default;
var prepareCustomQuestions_1 = require("./src/Functions/prepareCustomQuestions");
exports.prepareCustomQuestions = prepareCustomQuestions_1.default;
